src/app/centroidController.js

```diff
@@ -0,0 +1,255 @@
+              var deferreds = [];
+              var d = new Deferred();
+              var typesLookup = {
+                  0: 'poly-exploded',
+                  1: 'line-exploded',
+                  2: 'point-exploded'
+              };
+
+              this.centroidLayer = new FeatureLayer(config.urls.centroidService, {
+                  id: 'Centroid',
+                  mode: FeatureLayer.MODE_SELECTION,
+                  outFields: ['Title', 'Status', 'Project_ID']
+              });
+
+              this.centroidLayer.setVisibility(false);
+
+              this.centroidLayer.on('load', d.resolve);
+
+              deferreds.push(d);
+
+              [li.poly, li.line, li.point].forEach(function (layerIndex, i) {
+                  var layer = new FeatureLayer(config.urls.mapService + '/' + layerIndex, {
+                      id: typesLookup[i],
+                      mode: FeatureLayer.MODE_SELECTION
+                  });
+
+                  layer.setVisibility(false);
+
+                  var deferred = new Deferred();
+
+                  layer.on('load', deferred.resolve);
+
+                  deferreds.push(deferred);
+
+                  this.explodedLayer[typesLookup[i]] = layer;
+              }, this);
+
+              this.layersLoaded = true;
+
+              return all(deferreds);
+          }
+
+          return true;
+      },
+      updateLayerVisibilityFor: function (extent) {
+          // summary:
+          //      mimics the extent change event to trigger the decision tree of whether to show or hide centroids
+          // extent: {5:type or return: type}
+          console.log('app.mapControls.centroidController::updateLayerVisibilityFor', arguments);
+
+          if (!extent) {
+              return;
+          }
+
+          this.onExtentChanged(extent);
+      },
+      showFeaturesFor: function (where) {
+          // summary:
+          //      selects features for the where clause
+          // where: {5:type or return: type}
+          console.log('app.centroidController::showFeaturesFor', arguments);
+
+          var q = new Query();
+          q.where = where || this.where;
+
+          this.where = where;
+
+          var deferreds = [];
+
+          if (this.centroidsVisible) {
+              deferreds.push(this.centroidLayer.selectFeatures(q));
+          } else {
+              Object.keys(this.explodedLayer).forEach(function (key) {
+                  var layer = this.explodedLayer[key];
+                  deferreds.push(layer.selectFeatures(q));
+              }, this);
+          }
+
+          if (this.where === '1 = 1') {
+              return null;
+          }
+
+          return all(deferreds).then(
+              function (graphics) {
+                  if (graphics === null) {
+                      // state of utah extent
+                      return null;
+                  } else {
+                      return graphicsUtils.unionGraphicsIntoExtent(graphics);
```

Comment:
I think that this could be alternatively be done like this:
```
return esriGraphicsUtils.graphicsExtent(graphics.reduce(function (a, b) {
    return a.concat(b);
}));
```
which would eliminate the need for `unionGraphicsIntoExtent`.